default/AI/ShipDesignAI.py

```diff
@@ -0,0 +1,1573 @@
+
+            if verbose:
+                print "%s: " % universe.getPlanet(pid).name, self.parts_for_planets[pid]
+
+
+Cache = ShipDesignCache()
+
+
+class AdditionalSpecifications(object):
+    """This class is a container for all kind of additional information
+    and requirements we may want to use when assessing ship designs.
+
+    methods for external use:
+    convert_to_tuple(): Builds and returns a tuple of the class attributes
+    update_enemy(enemy): updates enemy stats
+    """
+
+    def __init__(self):
+        # TODO: Extend this framework according to needs of future implementations
+        self.minimum_fuel = 0
+        self.minimum_speed = 0
+        self.minimum_structure = 1
+        self.enemy_shields = 0
+        self.enemy_weapon_strength = 0
+        self.enemy_mine_dmg = 0  # TODO: Implement the detection of enemy mine damage
+        self.update_enemy(foAI.foAIstate.empire_standard_enemy)
+
+    def update_enemy(self, enemy):
+        """Read out the enemies stats and save them.
+
+        :param enemy: enemy as defined in AIstate
+        """
+        self.enemy_shields = enemy[2]
+        enemy_attack_stats = enemy[1]
+        self.enemy_weapon_strength = 0
+        for stat in enemy_attack_stats:
+            if stat[0] > self.enemy_weapon_strength:
+                self.enemy_weapon_strength = stat[0]
+
+    def convert_to_tuple(self):
+        """Create a tuple of this class' attributes (e.g. to use as key in dict).
+
+        :returns: tuple (minFuel,minSpeed,enemyDmg,enemyShield,enemyMineDmg)
+        """
+        tup = ("minFuel: %s" % self.minimum_fuel, "minSpeed: %s" % self.minimum_speed,
+               "enemyDmg: %s" % self.enemy_weapon_strength, "enemyShields: %s" % self.enemy_shields,
+               "enemyMineDmg: %s" % self.enemy_mine_dmg)
+        return tup
+
+
+class AIShipDesign(object):
+    """This class and its subclasses implement the building of a ship design and its rating.
+     Specialised Designs with their own rating system or optimizing algorithms should inherit from this class.
+
+    Member functions intended for external use:
+    optimize_design(): Returns the estimated optimum design according to the rating function
+    evaluate(): Returns a rating for the design as of the current state
+    update_hull(hullname): sets the hull used in the design
+    update_parts(partname_list): sets the parts used in the design
+    update_species(speciesname): sets the piloting species
+    update_stats(): calculates the stats of the design based on hull+parts+species
+    add_design(): Adds the shipdesign in the C++ part of the game
+
+    Functions which are to be overridden in inherited classes:
+    _rating_function()
+    _class_specific_filter()
+    _starting_guess()
+
+    For improved performance, maybe override _filling_algorithm() with a more specialised algorithm as well.
+    """
+    basename = "Default - Do not build"     # base design name
```

Comment:
It is static field that is not changed (but overridden in children) it is good practice to use UPPER_CASE for such names.